### ISSUE TITLE
DataCollector deadlock when using Test Impact Analysis (TIA) together with disabled CodeCoverage

### ISSUE THREAD

**Issue by Mertsch** [Microsoft/testfx#402]
## Description
See title

## Steps to reproduce
1. Have a multi core build machine
2. Run a UnitTest build with VSTest, MSTest 1.3.0-beta2, `[assembly: Parallelize(Scope = ExecutionScope.ClassLevel)]`, `Run only impacted tests` and `Run tests in parallel on multi-core machines` and the following `.runsettings`
```
<RunSettings>
  <RunConfiguration>
    <TargetPlatform>x64</TargetPlatform>
    <TargetFrameworkVersion>Framework45</TargetFrameworkVersion>
  </RunConfiguration>

  <DataCollectionRunSettings>
    <DataCollectors>
      <!-- https://docs.microsoft.com/en-us/visualstudio/test/customizing-code-coverage-analysis -->
      <DataCollector enabled="false" friendlyName="Code Coverage" uri="datacollector://Microsoft/CodeCoverage/2.0"
                     assemblyQualifiedName="Microsoft.VisualStudio.Coverage.DynamicCoverageDataCollector, Microsoft.VisualStudio.TraceCollector, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
        <Configuration>
          <CodeCoverage>
            <ModulePaths>
              <Exclude>
                <ModulePath>.*FakeItEasy.*</ModulePath>
                <ModulePath>.*GongSolutions.WPF.DragDrop.*</ModulePath>
              </Exclude>
            </ModulePaths>
            <Sources>
              <Exclude>
                <Source>.*\.Designer\.cs</Source>
              </Exclude>
            </Sources>
            <Attributes>
              <Exclude>
                <Attribute>^System\.Diagnostics\.CodeAnalysis.ExcludeFromCodeCoverageAttribute$</Attribute>
              </Exclude>
            </Attributes>
          </CodeCoverage>
        </Configuration>
      </DataCollector>
    </DataCollectors>
  </DataCollectionRunSettings>
</RunSettings>
```

## Expected behavior
The test process should exit and continue as usual.

## Actual behavior
`datacollector.exe` causes a deadlock and tests will never finish

Note: If you manually process-kill the `datacollector.exe`, the build will continue and finish

## Environment
* multi core build machine
* VS 15.6.6
* VSTest 15.6.2 or 15.7.0

**Comment by Mertsch** [comment]
I dug deeper and found that the disabled CC Collector causes the issue
`<DataCollector enabled="false" friendlyName="Code Coverage"`

**Comment by Mertsch** [comment]
In further Analysis it shows that even without CC the TIA data collector deadlocks ... The circumstances vary, but the effect is always the same.
Kill `datacollector.exe` and proceed...

**Comment by acesiddhu** [comment]
@Mertsch Sorry for the delay here. Can you take dump when datacollector.exe is hung and share it with us.

**Comment by Mertsch** [comment]
@acesiddhu Can you please tell me the CMD for that, I haven't taken a dump yet.

**Comment by acesiddhu** [comment]
@Mertsch you can take dumps from task manager. Goto details tab, right click on datacollector process and click take dumps option there

**Comment by Mertsch** [comment]
@acesiddhu Here is the download, please tell me when you got it
https://freedommanufaktur-my.sharepoint.com/:u:/g/personal/martin_freedom-manufaktur_com/EVcuj4VGjcxNr0__bMdQbLsBysjOpDGpZ2a_cnCWGw_5jg?e=khRcGK

**Comment by acesiddhu** [comment]
@Mertsch I got the dump files. will get back to you after investigations.

**Comment by mayerj** [comment]
I believe I have a similar, if not identical problem.  Here are the callstacks from datacollector.exe

0:004> !CLRStack
OS Thread Id: 0xb44 (4)
        Child SP               IP Call Site
000000f332aff0f8 00007ffbb572f0e4 [HelperMethodFrame_1OBJ: 000000f332aff0f8] System.Threading.WaitHandle.WaitOneNative(System.Runtime.InteropServices.SafeHandle, UInt32, Boolean, Boolean)
000000f332aff220 00007ffb9b4be98b System.Threading.WaitHandle.InternalWaitOne(System.Runtime.InteropServices.SafeHandle, Int64, Boolean, Boolean)
000000f332aff250 00007ffb9b4be95e System.Threading.WaitHandle.WaitOne(Int32, Boolean)
000000f332aff290 00007ffb3de69726 Microsoft.VisualStudio.Diagnostics.Logging.LoggerEngine.Stop(Boolean)
000000f332aff310 00007ffb3de66342 Microsoft.VisualStudio.TraceCollector.CommonDataCollector+LoggerWrapper.LoggedProcess_Exited(System.Object, Boolean)
000000f332aff360 00007ffb9b503423 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
000000f332aff430 00007ffb9b5032b4 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
000000f332aff460 00007ffb9beb1425 System.Threading._ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(System.Object, Boolean)
000000f332aff610 00007ffb9d406d93 [GCFrame: 000000f332aff610] 
000000f332aff808 00007ffb9d406d93 [DebuggerU2MCatchHandlerFrame: 000000f332aff808] 
0:004> ~5s
ntdll!NtWaitForMultipleObjects+0x14:
00007ffb`b572f0e4 c3              ret
0:005> !CLRStack
OS Thread Id: 0x51cc (5)
        Child SP               IP Call Site
000000f332cfe408 00007ffbb572f0e4 [GCFrame: 000000f332cfe408] 
000000f332cfe548 00007ffbb572f0e4 [GCFrame: 000000f332cfe548] 
000000f332cfe588 00007ffbb572f0e4 [HelperMethodFrame_1OBJ: 000000f332cfe588] System.Threading.Monitor.Enter(System.Object)
000000f332cfe680 00007ffb3de5d9fd Microsoft.VisualStudio.TraceCollector.CommonDataCollector+LoggerWrapper.SendMessage(MessageCreator, Boolean)
000000f332cfe730 00007ffb3de4d7ce Microsoft.VisualStudio.TraceCollector.CommonDataCollector.DispatchMessage(MessageCreator)
000000f332cfe7a0 00007ffb3de64dd8 Microsoft.VisualStudio.TraceCollector.CommonDataCollector.OnTestCaseEnd(System.Object, Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection.TestCaseEndEventArgs)
000000f332cfea98 00007ffb9d406d93 [DebuggerU2MCatchHandlerFrame: 000000f332cfea98] 
000000f332cfedd8 00007ffb9d406d93 [HelperMethodFrame_PROTECTOBJ: 000000f332cfedd8] System.RuntimeMethodHandle.InvokeMethod(System.Object, System.Object[], System.Signature, Boolean)
000000f332cfef50 00007ffb9b513e60 System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(System.Object, System.Object[], System.Object[])
000000f332cfefc0 00007ffb9b4e2519 System.Delegate.DynamicInvokeImpl(System.Object[])
000000f332cff010 00007ffb3de4b472 Microsoft.VisualStudio.TestPlatform.Utilities.MulticastDelegateUtilities.SafeInvoke(System.Delegate, System.Object, System.EventArgs, System.String)
000000f332cff070 00007ffb3de4aeea Microsoft.VisualStudio.TestPlatform.Common.DataCollector.TestPlatformDataCollectionEvents.RaiseEvent(Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection.DataCollectionEventArgs)
000000f332cff0c0 00007ffb3de648db Microsoft.VisualStudio.TestPlatform.Common.DataCollector.DataCollectionManager.TestCaseEnded(Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection.TestCaseEndEventArgs)
000000f332cff130 00007ffb3de5ee59 Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.DataCollection.DataCollectionTestCaseEventHandler.ProcessRequests()
000000f332cff180 00007ffb3de4e57d Microsoft.VisualStudio.TestPlatform.CommunicationUtilities.DataCollection.DataCollectionRequestHandler.b__22_0()
000000f332cff1e0 00007ffb9b54ca96 System.Threading.Tasks.Task.Execute()
000000f332cff220 00007ffb9b503423 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
000000f332cff2f0 00007ffb9b5032b4 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
000000f332cff320 00007ffb9b54cd35 System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)
000000f332cff3d0 00007ffb9b54c463 System.Threading.Tasks.Task.ExecuteEntry(Boolean)
000000f332cff410 00007ffb9b5425f2 System.Threading.ThreadPoolWorkQueue.Dispatch()
000000f332cff8a8 00007ffb9d406d93 [DebuggerU2MCatchHandlerFrame: 000000f332cff8a8]

**Comment by acesiddhu** [comment]
@mayerj the issue reported by @Mertsch is different than what you are seeing. if you are still seeing the issue please log a separate issue.
@Mertsch Is this consistent? the hang is actually while reading messages from socket.
Can you try with latest stable version of VS 15.9.0 once

**Comment by Mertsch** [comment]
@acesiddhu Yes it is consistent, I had to disable TIA on all my (10) pipelines and go with `TestCategory` to filter out long running tests.
Testing again is pretty annoying as I have to run TIA once and have it find tests the second time with appropriate checkins.

**Comment by acesiddhu** [comment]
@Mertsch unfortunately I am not able to repro this on my side and I feel the dump which tells a hang in socket is actually not a hang but it is waiting for message to come from the parent process. I am saying this because the stack show this is happening while TestRunStarted event which is before the test run starts. From the issue you reported it seems that hang happened after test run complete.
if you are not able to try on build definition is it possible for you to try on your dev machine where VS is installed? Use the runsetting which you shared earlier.

**Comment by Mertsch** [comment]
@acesiddhu So far it looks good with 15.9.6, I did not have too many builds yet, will get back to you in a week

**Comment by acesiddhu** [comment]
@Mertsch thanks for trying out. I will wait for your findings

**Comment by Mertsch** [comment]
@acesiddhu So far I could not observe a deadlock anymore. Let's hope it's fixed.
If I should see the behavior again, I will report back.
MSTest 1.4.0, Test SDK 15.9, VS 15.9.5
